### PR TITLE
Name arguments in Vm.sol

### PR DESCRIPTION
**Pull request by paulrberg**
Benefits:

1. Enable Forge Std users to call cheatcodes with [named arguments](https://twitter.com/PaulRBerg/status/1574071928544976896), which improves readability.
2. Better documentation for people who read the source code.

Examples:

```solidity
vm.roll({ newHeight: block.height + 100 });
vm.getCode({ path: "../../path/to/MyContract.sol" });
vm.expectEmit({ checkTopic1: true, checkTopic2: true, checkTopic3: true, checkData: true });
```

I think that this is especially useful for the `expectEmit` cheatcode, which, to quote the [Foundry Book](https://book.getfoundry.sh/forge/cheatcodes), is known for being "non-intuitive".

Notes:

- Generally I have tried to follow the argument names suggested in the comments
- When there was no name suggestion in the comments, I came up with my own (feedback welcome)

Questions:

1. Should we keep the argument names in the comments above the functions? They look a bit superfluous now.
2. I am torn between `txid` and `txhash`, which is needed in multiple functions. Which one do you like more?

### CODE REVIEW

**Comment by mds1** [comment]
Definitely better UX, thanks for this. I'll take a closer look at all the param names later. Also needs `forge fmt` 

> Should we keep the argument names in the comments above the functions? They look a bit superfluous now.

I'm indifferent here, I think I agree with removing them to make maintenance easier. Though we should probably also name return parameters too, which are currently missing

> I am torn between `txid` and `txhash`, which is needed in multiple functions. Which one do you like more?

I'd go with `txHash` personally, I think that's the most common

**Comment by paulrberg** [comment]
Thanks, glad you like it.

> Also needs `forge fmt`

Oh, shoot, I don't have it set up on VSCode yet. I applied `forge gmt` in my last commit.

> removing them to make maintenance easier.

Makes sense to me.

> name return parameters too, which are currently missing

Agree. I will try to do this later.

> I'd go with txHash personally, I think that's the most common

Good, changed in my last commit.

**Comment by paulrberg** [comment]
@mds1 I pushed another commit to name the return arguments.

While doing this, I realized that there is no explanation for the return argument of the `revertTo` cheatcode. I didn't know what to call it so I set as `result`, but I will let you decide if that makes sense or not based on the return value actually is.

File: src/Vm.sol
Comment (mds1) [inline src/Vm.sol:293]: [Seems](https://github.com/foundry-rs/foundry/blob/427c1b549be814b255d8a4dea8fa3c2409dddc4c/evm/src/executor/inspector/cheatcodes/snapshot.rs#L17) this returns `true` if you were able to revert to the given `snapshotId` and false otherwise, e.g. if it doesn't exist
```suggestion
    function revertTo(uint256 snapshotId) external returns (bool success);
```

**Comment by mds1** [comment]
There's also a few new cheatcodes now (such as https://github.com/foundry-rs/forge-std/pull/249) so needs conflicts resolved. @paulrberg if you want to add names for those that would be awesome, if not no problem and I can do it in a follow up PR

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol]: Thanks @mds1, fixed in last commit.

**Comment by paulrberg** [comment]
No worries @mds1, I refactored from master and fixed the git conflicts myself.

File: src/Vm.sol
Comment (mds1) [inline src/Vm.sol:61]: Do we want to give these a named return of `value` as we did with `envOr`?

**Review by mds1 — APPROVED** [review]
Looks great, found one missing a named return and once resolved we can merge!

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:61]: Yes. Actually they used to have the return value named, but I botched the rafactor.

Comment (paulrberg) [reply src/Vm.sol]: Fixed in the last commit.

**Review by mds1 — APPROVED** [review]
LGTM, thanks! Will wait for @ZeroEkkusu to take one more look before merging

**Comment by paulrberg** [comment]
I have just realized that the Foundry Book had already named the arguments, e.g. for [`createSelectFork`](https://book.getfoundry.sh/cheatcodes/create-select-fork):

Screenshot 2022-12-06 at 5 19 15 PM

Should we sync this PR with the Foundry Book?

I think it's @onbjerg who provided the names in the book.

**Comment by mds1** [comment]
Ah interesting! I'd say let's go with whichever is more descriptive/self-documenting for this PR, and we can always update the book to match

**Comment by paulrberg** [comment]
I might not have the time to go through each entry in the Foundry Book and update the args one-by-one.

I guess we could just this merge this PR as-is and update the arg names on the go?

**Comment by ZeroEkkusu** [comment]
Self-documentation > predictability ? @paulrberg

**Comment by ZeroEkkusu** [comment]
Thanks. I will review it for self-documentation first.

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:286]: Same as in Solidity (`block.chainid`).

```suggestion
    function chainId(uint256 newChainid) external;
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:288]:
```suggestion
    function store(address target, bytes32 slot, bytes32 value) external;
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:290]:
```suggestion
    function setNonce(address target, uint64 newNonce) external;
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:302]:
```suggestion
    function deal(address target, uint256 newBalance) external;
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:304]:
```suggestion
    function etch(address target, bytes calldata newRuntimeBytecode) external;
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:313]:
```suggestion
    function expectEmit(bool checkTopic1, bool checkTopic2, bool checkTopic3, bool checkData, address target)
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:319]:
```suggestion
    function mockCall(address target, bytes calldata data, bytes calldata returnData) external;
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:322]:
```suggestion
    function mockCall(address target, uint256 msgValue, bytes calldata data, bytes calldata returnData) external;
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:327]:
```suggestion
    function expectCall(address target, bytes calldata data) external;
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:329]:
```suggestion
    function expectCall(address target, uint256 msgValue, bytes calldata data) external;
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:341]:
```suggestion
    function createFork(string calldata url, uint256 blockNumber) external returns (uint256 forkId);
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:343]:
```suggestion
    function createFork(string calldata url) external returns (uint256 forkId);
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:346]:
```suggestion
    function createFork(string calldata url, bytes32 txHash) external returns (uint256 forkId);
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:348]:
```suggestion
    function createSelectFork(string calldata url, uint256 blockNumber) external returns (uint256 forkId);
```

File: src/Vm.sol
Comment (ZeroEkkusu) [inline src/Vm.sol:351]:
```suggestion
    function createSelectFork(string calldata url, bytes32 txHash) external returns (uint256 forkId);
```

**Review by ZeroEkkusu — COMMENTED** [review]
Sorry for the wait.

Good work.

Some suggestions: keep in mind I don't have a strong opinion about these - consider them, and discard all that aren't good.

File: src/Vm.sol
Comment (mds1) [reply src/Vm.sol:35]: Instead of target I'd suggest `contract` for the `load` and `store` cheats, since they aren't intended for use with EOAs

File: src/Vm.sol
Comment (mds1) [reply src/Vm.sol:41]: Personally prefer `account` here since it's standard nomenclature

File: src/Vm.sol
Comment (mds1) [reply src/Vm.sol:95]: `contract` instead of `target makes sense here for the same reason IMO

File: src/Vm.sol
Comment (mds1) [reply src/Vm.sol:35]: Oh right contract is a keyword, so maybe just `account`?

File: src/Vm.sol
Comment (mds1) [reply src/Vm.sol:95]: Same as https://github.com/foundry-rs/forge-std/pull/246#discussion_r1041668939

File: src/Vm.sol
Comment (mds1) [reply src/Vm.sol:101]: Maybe `account` again? `target` _feels_ a bit off though I don't have a great rationale 😅

File: src/Vm.sol
Comment (mds1) [reply src/Vm.sol:263]: The forge book/docs calls them aliases though, and I do think `alias` is more clear. The fact that we call it `key` in `StdChains` is an implementation detail that we don't need to leak to this interface, so I'd suggest sticking with `rpcAlias`

File: src/Vm.sol
Comment (mds1) [reply src/Vm.sol:286]: definitely a nit, though I prefer the `newChainId` since it's more readable and `chainId` style casing is very common elsewhere

File: src/Vm.sol
Comment (mds1) [reply src/Vm.sol:341]: `urlOrAlias` is probably best as @paulrberg pointed out above, since both are supported

**Comment by mds1** [comment]
Left a few replies to @ZeroEkkusu's suggestions, many I didn't reply to otherwise I'd end up leaving many duplicate comments. Overall I'm mostly indifferent on the suggestions fwiw as all of them are an improvement from the status quo

**Comment by ZeroEkkusu** [comment]
I’ve read @mds1 comments. @paulrberg, I’ll leave it to you to make any final changes, as both of us are indifferent on the suggestions.

And let’s get this merged.

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:41]: Yeah, I like `account` better in this context.

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:101]: I like `account` better in this context. It's EOAs who are usually labelled in tests.

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:263]: I went with `rpcAlias`.

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:286]: I went with `newChainId`.

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:288]: `target` is definitely a good fit in this context.

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:290]: I went with `account` here.

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:302]: I went with `account` here.

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:313]: I went with `emitter`, this is what's used in the Foundry Book. And I just like it better (nit).

File: src/Vm.sol
Comment (paulrberg) [reply src/Vm.sol:322]: I like `callee` better in this context because it refers to the action of calling a contract. `target` is much more broad.

**Review by ZeroEkkusu — APPROVED** [review]
Thank you. @mds1, you can merge if LG to you.

**Comment by mds1** [comment]
Thanks all!

**Pull request merged**
